### PR TITLE
=doc,htp #18657 document Headers/FormField request values

### PR DESCRIPTION
**Pull request by ktoso**
Season 2 of the Directives documentation bonanza 

Refs #18657

### CODE REVIEW

**Comment by jrudolph** [comment]
Error lines in https://jenkins.akka.io/job/pr-validator-per-commit-jenkins/3826/:

```
 13757 Test docs.http.javadsl.server.FormFieldRequestValsExampleTest.testFormFieldValsUnmarshaling failed: org.junit.ComparisonFailure: Unexpected entity! expected:<[Name: ..., age: ...]> but was:<[Request is missing required form field 'id']>, took 0.108 sec
 13758     at akka.http.javadsl.testkit.JUnitRouteTestBase$$anon$2.assertEquals(JUnitRouteTest.scala:28)
 13759     at akka.http.javadsl.testkit.TestResponse.assertEqualsKind(TestResponse.scala:160)
 13760     at akka.http.javadsl.testkit.TestResponse.assertEntity(TestResponse.scala:111)
 13761     at docs.http.javadsl.server.FormFieldRequestValsExampleTest.testFormFieldValsUnmarshaling(FormFieldRequestValsExampleTest.java:62)
 13762     ...
 13764 Test docs.http.javadsl.server.FormFieldRequestValsExampleTest.testFormFieldVals failed: org.junit.ComparisonFailure: Unexpected entity! expected:<[Name: ..., age: ...]> but was:<[Request is missing required form field 'name']>, took 0.033 sec
 13765     at akka.http.javadsl.testkit.JUnitRouteTestBase$$anon$2.assertEquals(JUnitRouteTest.scala:28)
 13766     at akka.http.javadsl.testkit.TestResponse.assertEqualsKind(TestResponse.scala:160)
 13767     at akka.http.javadsl.testkit.TestResponse.assertEntity(TestResponse.scala:111)
 13768     at docs.http.javadsl.server.FormFieldRequestValsExampleTest.testFormFieldVals(FormFieldRequestValsExampleTest.java:40)
 13769     ...
 16230 Failed: Total 271, Failed 2, Errors 0, Passed 269, Pending 12
 16231 Failed tests:
 16232 	docs.http.javadsl.server.FormFieldRequestValsExampleTest
 16283 (akka-docs-dev/test:test) sbt.TestsFailedException: Tests unsuccessful
 16284 (akka-docs-dev/sphinx:generatePdf) Sphinx pdf generation failed.  See debug output for details.
 16285 Total time: 773 s, completed Oct 7, 2015 6:10:46 PM
```

**Comment by ktoso** [comment]
Also resolves https://github.com/akka/akka/issues/18665 :)

File: akka-http-core/src/main/java/akka/http/javadsl/model/FormData.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import java.nio.charset.Charset;
+
+/**
+ * Simple model for `application/x-www-form-urlencoded` form data.
+ */
+public abstract class FormData {
+
+  protected abstract Query fields();
+
+  public RequestEntity toEntity() {
```
Comment (jrudolph) [inline akka-http-core/src/main/java/akka/http/javadsl/model/FormData.java]: Wouldn't it be nicer to implement it on the Scala side?

Comment (jrudolph) [inline akka-http-core/src/main/java/akka/http/javadsl/model/FormData.java]: Also, so that you can override the return value to the Scala RequestEntity?

**Comment by jrudolph** [comment]
Can you change the original FormData marshaller in akka-http as well to use the new method so that we don't introduce copies for the same functionality?

File: akka-http-core/src/main/java/akka/http/javadsl/model/headers/Host.java.orig [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    public static Host create(String host, int port) {
+      return akka.http.scaladsl.model.headers.Host.apply(host, port);
+    }
+
+>>>>>>> =doc,htp #18657 document Headers request values
+    public abstract akka.http.javadsl.model.Host host();
+    public abstract int port();
+}
```
Comment (ktoso) [inline akka-http-core/src/main/java/akka/http/javadsl/model/headers/Host.java.orig]: whoops

File: akka-http-core/src/main/java/akka/http/javadsl/model/FormData.java (hunk 1 above)
Comment (ktoso) [inline akka-http-core/src/main/java/akka/http/javadsl/model/FormData.java]: hm you're right, I forgot about Scala's `toEntity` will fix.

**Comment by jrudolph** [comment]
Error lines in https://jenkins.akka.io/job/pr-validator-per-commit-jenkins/3845/:

```
 16317 (akka-docs-dev/sphinx:generatePdf) Sphinx pdf generation failed.  See debug output for details.
 16318 Total time: 795 s, completed Oct 8, 2015 1:53:16 PM
```

**Comment by ktoso** [comment]
Due to compiler crazieness which took me a few hours today I did have to duplicate a bit of the code... 
The PredefinedMarshallers are using the ones defined here though.

File: akka-docs-dev/rst/java/code/docs/http/javadsl/server/FormFieldRequestValsExampleTest.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        .withEntity(formData.toEntity());
+    testRoute(route).run(request).assertEntity("Name: Blippy, age: 42");
+
+    //#simple
+  }
+
+  @Test
+  public void testFormFieldValsUnmarshaling() {
+    //#custom-unmarshal
+    FormField<SampleId> sampleId = FormFields.fromString("id", SampleId.class, s -> new SampleId(Integer.valueOf(s)));
```
Comment (patriknw) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/server/FormFieldRequestValsExampleTest.java]: too long line?

File: akka-docs-dev/rst/java/code/docs/http/javadsl/server/HeaderRequestValsExampleTest.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import akka.http.javadsl.testkit.JUnitRouteTest;
+import org.junit.Test;
+
+public class HeaderRequestValsExampleTest extends JUnitRouteTest {
+
+  @Test
+  public void testHeaderVals() {
+    //#by-class
+    // extract the entire header instance:
+    RequestVal<Host> host = Headers.byClass(Host.class).instance();
```
Comment (patriknw) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/server/HeaderRequestValsExampleTest.java]: is this the same as `RequestVals.host`? I guess this illustrates how to use `Headers`.

Comment (ktoso) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/server/HeaderRequestValsExampleTest.java]: Yes, does the same thing. I could pick a different header for the example I guess.
`Host` is the most prominent in all http requests so I picked that one.

File: akka-docs-dev/rst/java/code/docs/http/javadsl/server/HeaderRequestValsExampleTest.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+    final Route route =
+      route(
+        handleWith1(host, (ctx, h) ->
+          ctx.complete(String.format("Host header was: %s", h.host()))
+        )
+      );
+
+    // tests:
```
Comment (patriknw) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/server/HeaderRequestValsExampleTest.java]: perhaps we should have this kind of comment in all those code samples

Comment (ktoso) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/server/HeaderRequestValsExampleTest.java]: yes, good idea - I'll add it.
Makes it more obvious where tests start actually, can be surprising for java people I guess

File: akka-docs-dev/rst/java/http/routing-dsl/request-vals/form-field-request-vals.rst [hunk 6]
```diff
@@ -0,0 +1,25 @@
+.. _form-field-request-vals-java:
+
+Request Values: FormFields
+==========================
+
+A collection of pre-defined :ref:`request-vals-java` that can be used to extract header values from incoming requests.
+
+Description
```
Comment (patriknw) [inline akka-docs-dev/rst/java/http/routing-dsl/request-vals/form-field-request-vals.rst]: I found it rather redundant to have both the short description above and then this long description, so I removed this heading. I have no strong opinion but apart from making it consistent.

Comment (ktoso) [inline akka-docs-dev/rst/java/http/routing-dsl/request-vals/form-field-request-vals.rst]: Yeah, for some a "1 sentence summary" was nice but it gets repetitive really quick, I'll remove from all docs.

**Comment by patriknw** [comment]
LGTM

File: akka-docs-dev/rst/java/code/docs/http/javadsl/server/HeaderRequestValsExampleTest.java (hunk 4 above)
Comment (patriknw) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/server/HeaderRequestValsExampleTest.java]: no problem, I just wanted to confirm that I learnt something

Comment (ktoso) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/server/HeaderRequestValsExampleTest.java]: :)

File: akka-docs-dev/rst/java/code/docs/http/javadsl/server/HeaderRequestValsExampleTest.java (hunk 5 above)
Comment (ktoso) [inline akka-docs-dev/rst/java/code/docs/http/javadsl/server/HeaderRequestValsExampleTest.java]: done in the other PR

File: akka-docs-dev/rst/java/http/routing-dsl/request-vals/form-field-request-vals.rst (hunk 6 above)
Comment (ktoso) [inline akka-docs-dev/rst/java/http/routing-dsl/request-vals/form-field-request-vals.rst]: done in the other PR

**Pull request merged**